### PR TITLE
Various improvements to the xds proxy

### PR DESCRIPTION
**Pull request by howardjohn**
I have 6 commits here, happy to break them out, which is why its a WIP. The first is broken out into https://github.com/istio/istio/pull/27920.

Description of each change can be found in the commit messages

### CODE REVIEW

File: pkg/istio-agent/xds_proxy.go [hunk 1]
```diff
@@ -123,54 +128,97 @@ func initXdsProxy(ia *Agent) (*XdsProxy, error) {
 	if err = proxy.initCertificateWatches(ia, proxy.stopChan); err != nil {
 		return nil, err
 	}
+
+	go proxy.healthChecker.PerformApplicationHealthCheck(func(healthEvent *health.ProbeEvent) {
```
Comment (adiprerepa) [inline pkg/istio-agent/xds_proxy.go:132]: Curious - What is the upside to using a callback as opposed to directly using a channel here?

File: pkg/istio-agent/xds_proxy.go [hunk 2]
```diff
@@ -123,54 +128,97 @@ func initXdsProxy(ia *Agent) (*XdsProxy, error) {
...
+				con.downstreamError <- err
 				return
 			}
 			// forward to istiod
-			requestsChan <- req
+			con.requestsChan <- req
 			if !firstNDSSent && req.TypeUrl == v3.ListenerType {
 				// fire off an initial NDS request
-				ndsRequestChan <- &discovery.DiscoveryRequest{
+				con.requestsChan <- &discovery.DiscoveryRequest{
```
Comment (adiprerepa) [inline pkg/istio-agent/xds_proxy.go:211]: Whats the benefit to this as opposed to the `p.SendRequest(DiscoveryRequest)` method that was created it seems like for this purpose?

File: pkg/istio-agent/xds_proxy.go [hunk 3]
```diff
@@ -123,54 +128,97 @@ func initXdsProxy(ia *Agent) (*XdsProxy, error) {
...
+
+func (p *XdsProxy) RegisterStream(c *ProxyConnection) {
+	p.connectedMutex.Lock()
+	defer p.connectedMutex.Unlock()
+	if p.connected != nil {
+		close(p.connected.stopChan)
+	}
+	p.connected = c
+}
+
```
Comment (adiprerepa) [inline pkg/istio-agent/xds_proxy.go:169]: Why not create a constructor for ProxyConnection (`NewProxyConnection`) so we don't have to call `p.RegisterStream()` every time? Is there an instance where we wouldn't want the stream registered?

File: pkg/istio-agent/health/health_check_test.go [hunk 4]
```diff
@@ -15,127 +15,150 @@
...
+		if err != nil {
+			t.Errorf("unable to start mock listener: %v", err)
 		}
-	}
-	quitChan <- struct{}{}
+		httpServerEventCount := 0
+		srv := httptest.NewUnstartedServer(http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {
+			if httpServerEventCount < len(httpHealthStatuses) && httpHealthStatuses[httpServerEventCount] {
+				writer.WriteHeader(200)
+				writer.Write([]byte("foobar"))
```
Comment (adiprerepa) [inline pkg/istio-agent/health/health_check_test.go:137]: is this necessary? are we expecting a body?

**Review by adiprerepa — COMMENTED** [review]
Generally LGTM

File: pkg/istio-agent/xds_proxy.go (hunk 1 above)
Comment (howardjohn) [reply pkg/istio-agent/xds_proxy.go:132]: I was googling the same thing when I wrote this. I found https://www.jtolio.com/2016/03/go-channels-are-bad-and-you-should-feel-bad/ - see "Callbacks are strictly more powerful and don’t require unnecessary goroutines" - which I agree with. Basically, if you use a channel you are forcing the caller to use a certain pattern. callback is more flexible

File: pkg/istio-agent/health/health_check_test.go (hunk 4 above)
Comment (howardjohn) [reply pkg/istio-agent/health/health_check_test.go:137]: Its not necessary, it was just already there. I think its probably best to test with and without body if we want to be complete, although right now we don't.

File: pkg/istio-agent/xds_proxy.go (hunk 1 above)
Comment (adiprerepa) [reply pkg/istio-agent/xds_proxy.go:132]: good point. It also offers flexibility in tests from what it looks like.

**Comment by adiprerepa** [comment]
if #27920 is already broken out, why does it exist here at all?

**Comment by ramaraochavali** [comment]
Looks good at high level. Let me know when it is ready - I will take another look

**Comment by howardjohn** [comment]
I split out most of the PRs so this is quite a bit smaller now. I have been running it in my cluster for a while now without issues

File: pkg/istio-agent/xds_proxy.go [hunk 5]
```diff
@@ -124,69 +129,116 @@ func initXdsProxy(ia *Agent) (*XdsProxy, error) {
...
+type ProxyConnection struct {
```
Comment (hzxuzhonghu) [inline pkg/istio-agent/xds_proxy.go:171]: Where is the ndsRequestChan

File: pkg/istio-agent/xds_proxy.go [hunk 6]
```diff
@@ -124,69 +129,116 @@ func initXdsProxy(ia *Agent) (*XdsProxy, error) {
...
+	}
 
-	upstreamError := make(chan error)
-	downstreamError := make(chan error)
-	requestsChan := make(chan *discovery.DiscoveryRequest, 10)
-	responsesChan := make(chan *discovery.DiscoveryResponse, 10)
-	healthEventsChan := make(chan *health.ProbeEvent, 5)
-	// A separate channel for nds requests to not contend with the ones from envoys
-	ndsRequestChan := make(chan *discovery.DiscoveryRequest, 5)
+	p.RegisterStream(con)
```
Comment (hzxuzhonghu) [inline pkg/istio-agent/xds_proxy.go:195]: I have the similar doubt with @adiprerepa 

Why donot put the proxyConnection initialization in `initXdsProxy`, and i can not see why `downstream` is necessary. 

And furthermore we can eliminate the lock

File: pkg/istio-agent/xds_proxy.go [hunk 7]
```diff
@@ -124,69 +129,116 @@ func initXdsProxy(ia *Agent) (*XdsProxy, error) {
...
 
-	// health check stop channel
-	stop := make(chan struct{})
-	defer close(stop)
-	go p.healthChecker.PerformApplicationHealthCheck(healthEventsChan, stop)
-
-	upstreamConn, err := grpc.Dial(p.istiodAddress, p.istiodDialOptions...)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+	upstreamConn, err := grpc.DialContext(ctx, p.istiodAddress, p.istiodDialOptions...)
```
Comment (hzxuzhonghu) [inline pkg/istio-agent/xds_proxy.go:221]: This still couples the downstream with the upstream

Take a look at https://github.com/istio/istio/pull/27884, we can move it out

File: pkg/istio-agent/xds_proxy.go (hunk 2 above)
Comment (howardjohn) [reply pkg/istio-agent/xds_proxy.go:211]: Its likely the same in 99.9% of cases. Some benefit of this is that it guarantees its not sending the request to a different stream in some edge cases where there are two streams for a short period of time, and avoids locking.

File: pkg/istio-agent/xds_proxy.go (hunk 5 above)
Comment (howardjohn) [reply pkg/istio-agent/xds_proxy.go:171]: I do not think we need a dedicated channel for each type. The reasons we had it initially are no longer valid as we have a buffered channel now.

File: pkg/istio-agent/xds_proxy.go (hunk 6 above)
Comment (howardjohn) [reply pkg/istio-agent/xds_proxy.go:195]: initXdsProxy initializes a singleton server. This server will have multiple streams (calls to StreamAggregatedResources) established, of which `ProxyConnection` is local to.

File: pkg/istio-agent/xds_proxy.go (hunk 7 above)
Comment (howardjohn) [reply pkg/istio-agent/xds_proxy.go:221]: This is intentional. If you do not couple them, you break the statefulness of upstream and downstream. For example, if upstream disconnects and we reconnect to upstream, then we will not send the full ADS initialization sequence (CDS req, etc), and we will not get updates. Upstream:Downstream must be 1:1

File: pkg/istio-agent/xds_proxy.go (hunk 5 above)
Comment (hzxuzhonghu) [reply pkg/istio-agent/xds_proxy.go:171]: make sense, just make sure no missing

File: pkg/istio-agent/xds_proxy.go (hunk 7 above)
Comment (hzxuzhonghu) [reply pkg/istio-agent/xds_proxy.go:221]: It make sense in this initial sequence, but i think lose the initial intention of xds proxy, which is to save connections to istiod

Comment (howardjohn) [reply pkg/istio-agent/xds_proxy.go:221]: The original intent was to have a single connection (before we had one from envoy for XDS and one from agent for CSR) not to to reduce the number of new connections although that is something we can do. I think I was mixing up connection with stream a bit. We MUST be 1:1 on the stream, but maybe we can reuse the connection. We will need to do more research on what happens when istiod closes the connection though if we break it out like https://github.com/istio/istio/pull/27884/files

Comment (howardjohn) [reply pkg/istio-agent/xds_proxy.go:221]: btw, isn't it the same number of connections most of the time? Envoy will never close the stream, so the only time we will reconnect is if the upstream has an error, in which case we need to reconnect anyways. It seems like mostly a refactoring to call `grpc.Dial` once and expect it to continuously reconnect automatically (as done in https://github.com/istio/istio/pull/27884/files) or to directly call `grpc.Dial` as in this PR?

Comment (hzxuzhonghu) [reply pkg/istio-agent/xds_proxy.go:221]: It can be a follow-up.

**Review by hzxuzhonghu — APPROVED** [review]

**Comment by hzxuzhonghu** [comment]
@howardjohn Should we cherrypick this to 1.8?

**Comment by nrjpoddar** [comment]
@howardjohn i see that there’s a unit test related to the bug fix which is great, but do you think an integration test to verify timeout behavior would help?

**Pull request merged**